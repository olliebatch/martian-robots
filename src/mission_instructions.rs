use crate::robots::{Robot, RobotPosition};
use anyhow::anyhow;
use std::fmt;
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub struct Coordinates {
    pub x: i32,
    pub y: i32,
}

impl FromStr for Coordinates {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // remove any spaces if necessary
        let coords: String = s.chars().filter(|c| !c.is_whitespace()).collect();

        if coords.len() != 2 {
            //handle nicer error
            return Err(anyhow!("Too many characters for Coordinates"));
        }
        let mut chars = coords.chars();
        let x_from_char = chars.next().unwrap().to_string().parse::<i32>()?;
        let y_from_char = chars.next().unwrap().to_string().parse::<i32>()?;

        Ok(Coordinates {
            x: x_from_char,
            y: y_from_char,
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Command {
    pub upper_right: Coordinates,
    pub robots: Vec<Robot>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Orientation {
    North,
    South,
    West,
    East,
}

impl fmt::Display for Orientation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let val = match self {
            Orientation::North => "N",
            Orientation::South => "S",
            Orientation::West => "W",
            Orientation::East => "E",
        };
        write!(f, "{:}", val)
    }
}

impl FromStr for Orientation {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        match input {
            "N" => Ok(Orientation::North),
            "S" => Ok(Orientation::South),
            "W" => Ok(Orientation::West),
            "E" => Ok(Orientation::East),
            _ => Err(anyhow!("Error matching orientation")),
        }
    }
}
impl Orientation {
    pub fn change_right(&self) -> Self {
        match self {
            Orientation::North => Orientation::East,
            Orientation::East => Orientation::South,
            Orientation::South => Orientation::West,
            Orientation::West => Orientation::North,
        }
    }
    pub fn change_left(&self) -> Self {
        match self {
            Orientation::North => Orientation::West,
            Orientation::East => Orientation::North,
            Orientation::South => Orientation::East,
            Orientation::West => Orientation::South,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum RobotCommands {
    Left,
    Right,
    Forward,
}

impl RobotCommands {
    pub fn process(&self, robot_position: RobotPosition) -> RobotPosition {
        match self {
            RobotCommands::Right => {
                let new_orientation = robot_position.orientation.change_right();
                robot_position.update_orientation(new_orientation)
            }
            RobotCommands::Left => {
                let new_orientation = robot_position.orientation.change_left();
                robot_position.update_orientation(new_orientation)
            }
            RobotCommands::Forward => robot_position.move_forward(),
        }
    }
}

impl FromStr for RobotCommands {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        match input {
            "L" => Ok(RobotCommands::Left),
            "R" => Ok(RobotCommands::Right),
            "F" => Ok(RobotCommands::Forward),
            _ => Err(anyhow!("Error matching possible Robot Commands")),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
    use std::str::FromStr;

    use rstest::*;

    #[rstest]
    #[case("1 2", 1, 2)]
    #[case("12", 1, 2)]
    #[case("2 3", 2, 3)]
    #[case("23", 2, 3)]
    fn test_from_str_for_coords(
        #[case] input: &str,
        #[case] expected_x: i32,
        #[case] expected_y: i32,
    ) {
        let p = Coordinates::from_str(input);
        assert_eq!(
            p.unwrap(),
            Coordinates {
                x: expected_x,
                y: expected_y
            }
        )
    }

    #[rstest]
    #[case("N", Orientation::North)]
    #[case("W", Orientation::West)]
    #[case("E", Orientation::East)]
    #[case("S", Orientation::South)]
    fn test_orientation_from_str(#[case] input: &str, #[case] expected_orientation: Orientation) {
        let p = Orientation::from_str(input);
        assert_eq!(p.unwrap(), expected_orientation)
    }

    #[test]
    fn test_orientation_err_from_str() {
        let p = Orientation::from_str("Y");
        assert!(p.is_err())
    }

    #[rstest]
    #[case("F", RobotCommands::Forward)]
    #[case("L", RobotCommands::Left)]
    #[case("R", RobotCommands::Right)]
    fn test_robot_commands(#[case] input: &str, #[case] expected_command: RobotCommands) {
        let command = RobotCommands::from_str(input);

        assert_eq!(command.unwrap(), expected_command);
    }

    #[rstest]
    #[case(Orientation::North, Orientation::East)]
    #[case(Orientation::South, Orientation::West)]
    #[case(Orientation::East, Orientation::South)]
    #[case(Orientation::West, Orientation::North)]
    fn test_moving_right(#[case] input: Orientation, #[case] expected_orientation: Orientation) {
        let p = input.change_right();
        assert_eq!(p, expected_orientation)
    }
    #[rstest]
    #[case(Orientation::North, Orientation::West)]
    #[case(Orientation::South, Orientation::East)]
    #[case(Orientation::East, Orientation::North)]
    #[case(Orientation::West, Orientation::South)]
    fn test_moving_left(#[case] input: Orientation, #[case] expected_orientation: Orientation) {
        let p = input.change_left();
        assert_eq!(p, expected_orientation)
    }
}
