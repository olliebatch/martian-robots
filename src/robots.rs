use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub struct RobotPosition {
    pub coordinates: Coordinates,
    pub orientation: Orientation,
}

impl RobotPosition {
    pub fn update_orientation(self, orientation: Orientation) -> RobotPosition {
        RobotPosition {
            coordinates: self.coordinates,
            orientation,
        }
    }

    pub fn move_forward(&self) -> Self {
        let new_coordinates = match self.orientation {
            Orientation::North => Coordinates {
                x: self.coordinates.x,
                y: self.coordinates.y + 1,
            },
            Orientation::South => Coordinates {
                x: self.coordinates.x,
                y: self.coordinates.y - 1,
            },
            Orientation::East => Coordinates {
                x: self.coordinates.x + 1,
                y: self.coordinates.y,
            },
            Orientation::West => Coordinates {
                x: self.coordinates.x - 1,
                y: self.coordinates.y,
            },
        };
        RobotPosition {
            coordinates: new_coordinates,
            orientation: self.orientation.to_owned(),
        }
    }
}

impl fmt::Display for RobotPosition {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{:} {:} {:}",
            self.coordinates.x, self.coordinates.y, self.orientation
        )
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Robot {
    pub position: RobotPosition,
    pub robot_commands: Vec<RobotCommands>,
}

impl Robot {
    #[cfg(test)]
    pub fn new_basic_robot() -> Self {
        let robot_position = RobotPosition {
            coordinates: Coordinates { x: 1, y: 1 },
            orientation: Orientation::East,
        };
        let robot_commands = vec![
            RobotCommands::Right,
            RobotCommands::Forward,
            RobotCommands::Right,
            RobotCommands::Forward,
            RobotCommands::Right,
            RobotCommands::Forward,
            RobotCommands::Right,
            RobotCommands::Forward,
        ];
        Robot {
            robot_commands,
            position: robot_position,
        }
    }

    pub fn new() -> Self {
        Robot {
            position: RobotPosition {
                coordinates: Coordinates { x: 0, y: 0 },
                orientation: Orientation::North,
            },
            robot_commands: vec![],
        }
    }

    pub fn set_start_position(self, robot_position: RobotPosition) -> Self {
        Robot {
            position: robot_position,
            robot_commands: self.robot_commands,
        }
    }

    pub fn update_commands(self, robot_commands: Vec<RobotCommands>) -> Self {
        Robot {
            position: self.position,
            robot_commands,
        }
    }

    pub fn process_robot_command(mut self, coordinate_limit: &Coordinates) -> Self {
        let command = self.robot_commands.first().unwrap();
        let new_position = command.process(self.position, coordinate_limit);
        self.robot_commands.remove(0);
        Robot {
            position: new_position,
            robot_commands: self.robot_commands,
        }
    }

    pub fn process_all_commands(mut self, coordinate_limit: &Coordinates) -> Self {
        while !self.robot_commands.is_empty() {
            self = self.process_robot_command(coordinate_limit)
        }
        self
    }
}

#[cfg(test)]
mod test {

    use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
    use crate::robots::{Robot, RobotPosition};
    use rstest::*;

    #[rstest]
    #[case(Orientation::North)]
    #[case(Orientation::South)]
    #[case(Orientation::West)]
    #[case(Orientation::East)]
    fn test_from_str_for_coords(#[case] orientation: Orientation) {
        let robot_position = RobotPosition {
            coordinates: Coordinates { x: 2, y: 2 },
            orientation,
        };

        let moved = robot_position.move_forward();
        insta::assert_debug_snapshot!(moved)
    }

    #[rstest]
    #[case(RobotCommands::Forward)]
    #[case(RobotCommands::Left)]
    #[case(RobotCommands::Right)]

    fn test_single_rob_command(#[case] command: RobotCommands) {
        let robot = Robot::new();
        let robot_with_commands = robot.update_commands(vec![command]);
        let processed_robot = robot_with_commands.process_robot_command();
        insta::assert_debug_snapshot!(processed_robot)
    }

    #[test]
    fn test_process_all_commands() {
        let new_basic_robot = Robot::new_basic_robot();
        let processed_robot = new_basic_robot.process_all_commands();
        insta::assert_debug_snapshot!(processed_robot)
    }
}
