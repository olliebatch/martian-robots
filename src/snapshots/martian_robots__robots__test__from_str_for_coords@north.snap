---
source: src/robots.rs
assertion_line: 168
expression: moved
---
RobotPosition {
    coordinates: Coordinates {
        x: 2,
        y: 3,
    },
    orientation: North,
}
